### PR TITLE
DEMO-115 : getFollowers, getFollowing, fix profileDto

### PR DESCRIPTION
**Pull request by sseerrnn**
- Add getFollowers
- Add getFollowing
- Split getProfile in to getMyProfile and getUserProfile
- Add roles in profileDto

### CODE REVIEW

**Comment by sseerrnn** [comment]
ProfileDto
-   followerCount: number;
-   followingCount: number;
-   playlistCount: number;
-   profileImage: string;
-   username: string;
-   roles: Role[];

File: src/users/users.controller.ts [hunk 1]
```diff
@@ -221,16 +221,36 @@ export class UsersController {
...
     description: 'Return user profile',
   })
-  @Get('profile')
+  @Get('profile/me')
   @HttpCode(HttpStatus.OK)
-  async getUserProfile(@Req() req): Promise<ProfileDto> {
+  async getMyProfile(@Req() req): Promise<ProfileDto> {
     const user = await this.usersService.findOneByEmail(req.user.email);
+    const follower = await this.usersService.getFollowers(user.username);
+    const profile: ProfileDto = new ProfileDto();
```
Comment (pol-alones) [inline src/users/users.controller.ts]: You should create this part in the user service. Also you can write the code as
```
const profile: ProfileDto = {
followerCount: follower.length,
.
.
.
}
```

File: src/users/users.controller.ts [hunk 2]
```diff
@@ -221,16 +221,36 @@ export class UsersController {
...
+    profile.roles = user.roles;
+
+    return profile;
+  }
+
+  @Get('profile/:id')
+  @ApiParam({ name: 'id' })
+  @HttpCode(HttpStatus.OK)
+  async getUserProfile(@Req() req): Promise<ProfileDto> {
+    const user = await this.usersService.findOneById(req.params.id);
```
Comment (pol-alones) [inline src/users/users.controller.ts]: Create a function in user service

File: src/users/users.controller.ts [hunk 3]
```diff
@@ -262,4 +282,20 @@ export class UsersController {
     await this.usersService.unfollowArtist(req.user.email, params.followeeName);
     return { message: 'success to unfollow user', success: true };
   }
+
+  @ApiBearerAuth()
```
Comment (pol-alones) [inline src/users/users.controller.ts]: This decorator is unnecessary since I declared it in global.

File: src/users/users.controller.ts [hunk 4]
```diff
@@ -221,16 +221,36 @@ export class UsersController {
   @ApiOkResponse({
     description: 'Return user profile',
   })
```
Comment (pol-alones) [inline src/users/users.controller.ts:229]: API Response type?

File: src/users/users.controller.ts [hunk 5]
```diff
@@ -221,16 +221,36 @@ export class UsersController {
...
+    profile.followingCount = user.following.length;
+    profile.playlistCount = 999;
+    profile.username = user.username;
+    profile.profileImage = user.profileImage;
+    profile.roles = user.roles;
+
+    return profile;
+  }
+
+  @Get('profile/:id')
```
Comment (pol-alones) [inline src/users/users.controller.ts:241]: Response type?

File: src/users/users.controller.ts [hunk 6]
```diff
@@ -262,4 +282,20 @@ export class UsersController {
     await this.usersService.unfollowArtist(req.user.email, params.followeeName);
     return { message: 'success to unfollow user', success: true };
   }
+
+  @ApiBearerAuth()
+  @ApiParam({ name: 'followeeName' })
+  @Get('follower/:followeeName')
+  @HttpCode(HttpStatus.OK)
```
Comment (pol-alones) [inline src/users/users.controller.ts:298]: Response type?

File: src/users/users.controller.ts [hunk 7]
```diff
@@ -262,4 +282,20 @@ export class UsersController {
...
+  async getFollowers(@Param() params) {
+    return await this.usersService.getFollowers(params.followeeName);
+  }
+
+  @ApiBearerAuth()
```
Comment (pol-alones) [inline src/users/users.controller.ts]: Response type?

File: src/users/users.service.ts [hunk 8]
```diff
@@ -174,4 +174,61 @@ export class UsersService {
...
   }
+
+  async getFollowers(username: string): Promise<UserDto[]> {
+    const user = await this.findOneByUsername(username);
+    if (!user) {
+      throw new NotFoundException(
+        `There isn't any user with username: ${username}`,
+      );
+    }
+    const followers = await this.userModel.aggregate([
```
Comment (pol-alones) [inline src/users/users.service.ts]: The aggregation you are doing is very inefficient.

File: src/users/users.service.ts [hunk 9]
```diff
@@ -174,4 +174,61 @@ export class UsersService {
...
+      {
+        $project: {
+          _id: 1,
+          username: 1,
+          email: 1,
+          profileImage: 1,
+        },
+      },
+    ]);
+    return followers;
```
Comment (pol-alones) [inline src/users/users.service.ts:211]: Is it in format of UserDto?

File: src/users/users.service.ts [hunk 10]
```diff
@@ -174,4 +174,61 @@ export class UsersService {
...
+      {
+        $project: {
+          _id: 1,
+          username: 1,
+          email: 1,
+          profileImage: 1,
+        },
+      },
+    ]);
+    return following;
```
Comment (pol-alones) [inline src/users/users.service.ts:251]: Is it in format of UserDto?

File: src/users/users.service.ts (hunk 8 above)
Comment (pol-alones) [reply src/users/users.service.ts]: try `.find()` with filter that the userId is in `following` of others

File: src/users/users.controller.ts (hunk 3 above)
Comment (sseerrnn) [reply src/users/users.controller.ts]: Can I delete all of its?

**Pull request merged**